File: HW/app/backend/app/models/product.py
"""Product data models for the online store API."""

from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

# Define valid product categories as a type alias for reusability
ProductCategory = Literal["electronics", "clothing", "home", "sports", "books"]


class Product(BaseModel):
    """
    Represents a single product in the online store.

    Attributes:
        product_id: Unique identifier for the product
        product_name: Display name of the product
        product_description: Detailed description of the product
        product_price_usd: Price in US dollars (uses Decimal for precision)
        product_category: One of the predefined product categories
        product_in_stock: Whether the product is currently available
    """

    product_id: int = Field(..., description="Unique product identifier", gt=0, examples=[1, 42, 1337])

    product_name: str = Field(
        ...,
        description="Display name of the product",
        min_length=1,
        max_length=200,
        examples=["Wireless Mouse", "Cotton T-Shirt", "Coffee Maker"],
    )

    product_description: str = Field(
        ...,
        description="Detailed description of the product features and specifications",
        min_length=1,
        max_length=1000,
        examples=["Ergonomic wireless mouse with USB receiver and long battery life"],
    )

    product_price_usd: Decimal = Field(
        ...,
        description="Product price in US dollars (uses Decimal for monetary precision)",
        gt=0,
        decimal_places=2,
        examples=["29.99", "199.99", "9.99"],
    )

    product_category: ProductCategory = Field(
        ...,
        description="Product category, must be one of the predefined categories",
        examples=["electronics", "clothing", "home", "sports", "books"],
    )

    product_in_stock: bool = Field(default=True, description="Whether the product is currently available for purchase")


class ProductListResponse(BaseModel):
    """
    Response model for endpoints that return a list of products.

    Attributes:
        products: List of product objects
        total_count: Number of products in the response
    """

    products: list[Product] = Field(..., description="List of products matching the request criteria")

    total_count: int = Field(..., description="Total number of products in this response", ge=0)

File: HW/app/frontend/src/types/product.ts
/**
 * Product types matching backend Pydantic models EXACTLY.
 *
 * Backend definitions (app/models/product.py):
 * - ProductCategory = Literal["electronics", "clothing", "home", "sports", "books"]
 * - Product model with strict field types
 * - ProductListResponse wrapper
 *
 * IMPORTANT: These types must stay in sync with backend models.
 */

/**
 * Product category enum matching backend Literal type.
 *
 * Backend: ProductCategory = Literal["electronics", "clothing", "home", "sports", "books"]
 */
export type ProductCategory = "electronics" | "clothing" | "home" | "sports" | "books";

/**
 * Product model matching backend Pydantic Product model.
 *
 * Note: Decimal from backend is serialized as string in JSON
 */
export interface Product {
  /** Unique product identifier (always positive integer) */
  product_id: number;

  /** Display name of the product (1-200 characters) */
  product_name: string;

  /** Detailed product description (1-1000 characters) */
  product_description: string;

  /** Price in USD as string (Decimal serialized from backend with 2 decimal places) */
  product_price_usd: string;

  /** Product category (one of 5 valid categories) */
  product_category: ProductCategory;

  /** Whether product is currently available for purchase */
  product_in_stock: boolean;
}

/**
 * Product list response matching backend ProductListResponse model.
 */
export interface ProductListResponse {
  /** Array of product objects */
  products: Product[];

  /** Total number of products in response (always >= 0) */
  total_count: number;
}
